src/flb_upstream.c

```diff
@@ -191,14 +191,14 @@ static int destroy_conn(struct flb_upstream_conn *u_conn)
 #endif
 
     if (u_conn->fd > 0) {
+        u_conn->fd = -1;
```

Comment:
The PR description says:
 > In the destroy connection, the connection object is removed from the list twice and the object is freed twice. The fix, sets the socket fd to "-1" after the first destruction and doesn't take any action if the destroy is called the 2nd time.

But, in this function, we call flb_free on `u_conn`. So if `destroy_conn` is being called twice, then this doesn't help. 

The check will throw a sigsev:

```
if (u_conn->fd > 0) {
```

If u_conn has been freed. Which it will after this function runs.

Reply:
@PettitWesley 
Please check the below logs. (I have changed the log level to debug from trace since I had not enable trace). This is without fd being set to -1.

```
 34 ^[[1m[^[[0m2020/08/28 10:59:21^[[1m]^[[0m [^[[93mdebug^[[0m] [upstream] KA connection #16 is in a failed st    ate to: 192.168.99.99:80, cleaning up
 35 ^[[1m[^[[0m2020/08/28 10:59:21^[[1m]^[[0m [^[[93mdebug^[[0m] [upstream] destroy connection #16 to 192.168.9    9.99:80
 36 ^[[1m[^[[0m2020/08/28 10:59:59^[[1m]^[[0m [^[[91merror^[[0m] [http_client] broken connection to 192.168.99.    99:80 ?
 37 ^[[1m[^[[0m2020/08/28 10:59:59^[[1m]^[[0m [^[[91merror^[[0m] [output:http:http.0] could not flush records t    o 192.168.99.99:80 (http_do=-1)
 38 ^[[1m[^[[0m2020/08/28 10:59:59^[[1m]^[[0m [^[[93mdebug^[[0m] [upstream] KA connection #16 to 192.168.99.99:    80 is now available
 39 ^[[1m[^[[0m2020/08/28 10:59:59^[[1m]^[[0m [^[[93mdebug^[[0m] [upstream] KA connection #16 to 192.168.99.99:    80 has been disconnected by the remote service
 40 ^[[1m[^[[0m2020/08/28 10:59:59^[[1m]^[[0m [^[[93mdebug^[[0m] [upstream] destroy connection #16 to 192.168.9    9.99:80

```

We can see that for connection 16 destroy is being called twice.

The first caller is 
src/flb_upstream.c
```
struct flb_upstream_conn *flb_upstream_conn_get(struct flb_upstream *u)
.
.
.
345         int err;
346         err = flb_socket_error(conn->fd);
347         if (!FLB_EINPROGRESS(err)) {
348             flb_debug("[upstream] KA connection #%i is in a failed state "
349                       "to: %s:%i, cleaning up",
350                       conn->fd, u->tcp_host, u->tcp_port);
351             destroy_conn(conn);
352             conn = NULL;
353             continue;
354         }
```
The second caller is:
src/flb_upstream.c
```
385 static int cb_upstream_conn_ka_dropped(void *data)
386 {
387     struct flb_upstream_conn *conn;
388
389     conn = (struct flb_upstream_conn *) data;
390
391     flb_debug("[upstream] KA connection #%i to %s:%i has been disconnected "
392               "by the remote service",
393               conn->fd, conn->u->tcp_host, conn->u->tcp_port);
394     return destroy_conn(conn);
395 }

```

The fix might not be a proper one, but I think the root cause is.

I am also attaching the valgrind output. You can check the below piece of log in the file.
==17924== Invalid write of size 8
==17924==    at 0x1745F1: __mk_list_del (in /usr/local/bin/fluent-bit)
==17924==    by 0x174628: mk_list_del (in /usr/local/bin/fluent-bit)
==17924==    by 0x174F7B: destroy_conn (in /usr/local/bin/fluent-bit)
==17924==    by 0x175581: cb_upstream_conn_ka_dropped (in /usr/local/bin/fluent-bit)
==17924==    by 0x16E712: flb_engine_start (in /usr/local/bin/fluent-bit)
==17924==    by 0x14FFD3: flb_main (in /usr/local/bin/fluent-bit)
==17924==    by 0x150021: main (in /usr/local/bin/fluent-bit)
==17924==  Address 0x8 is not stack'd, malloc'd or (recently) free'd
==17924==

Reply:
After digging a little bit more, found that after the first deletion the event registered should have been removed and the callback (cb_upstream_conn_ka_dropped) shouldn't be called.
The code for deletion seems to be there but doesn't seem to be remove the registered event.

In destroy_conn I see call to mk_event_del(u->evl, &u_conn->event);

Reply:
@phiremande What output are you using?

I wonder if the bug is:

```
if (u->flags & FLB_IO_ASYNC) {
        mk_event_del(u->evl, &u_conn->event);
    }
```

The code that adds the event does not check for the async flag. So may be this is happens when async is not enabled, so the event is added, but then not removed in destroy_conn.

If you tell me what plugins you are using, we can figure out if you are running anything in sync mode or not. Most plugins run in sync mode, except for a few.

Reply:
@PettitWesley I am using http output. Btw, I had even tried adding logs and capturing the results. The result of the call mk_event_del is '0', which indicates it is success.

I am attaching the logs and the diff for your reference so that you can track from the additional logs added. Unless I am obviously missing something, I see that the mk_event_del is getting called and the first time the result is '0' (success). After that only the event is again triggered form the deleted fd.

pasting the diff and logs here as attaching files isn't working for some reason.

```
diff --git a/lib/monkey/include/monkey/mk_core/mk_event.h b/lib/monkey/include/monkey/mk_core/mk_event.h
index 4860ab29..ad61d2b6 100644
--- a/lib/monkey/include/monkey/mk_core/mk_event.h
+++ b/lib/monkey/include/monkey/mk_core/mk_event.h
@@ -98,6 +98,7 @@ struct mk_event_loop {
     int n_events;              /* number of events reported */
     struct mk_event *events;   /* copy or reference of events triggered */
     void *data;                /* mk_event_ctx_t from backend */
+
 };
 
 static inline void MK_EVENT_INIT(struct mk_event *ev, int fd, void *data,
diff --git a/lib/monkey/mk_core/mk_event_epoll.c b/lib/monkey/mk_core/mk_event_epoll.c
index ab072f45..6af53b66 100644
--- a/lib/monkey/mk_core/mk_event_epoll.c
+++ b/lib/monkey/mk_core/mk_event_epoll.c
@@ -145,6 +145,7 @@ static inline int _mk_event_add(struct mk_event_ctx *ctx, int fd,
 static inline int _mk_event_del(struct mk_event_ctx *ctx, struct mk_event *event)
 {
     int ret;
+    printf("PHK MK deleting %i\n", event->fd);
 
     ret = epoll_ctl(ctx->efd, EPOLL_CTL_DEL, event->fd, NULL);
     MK_TRACE("[FD %i] Epoll, remove from QUEUE_FD=%i, ret=%i",
@@ -155,6 +156,7 @@ static inline int _mk_event_del(struct mk_event_ctx *ctx, struct mk_event *event
 #endif
     }
 
+    printf("PHK MK deleted %i\n", event->fd);
     return ret;
 }
 
diff --git a/src/flb_engine.c b/src/flb_engine.c
index efd134e3..e042b2ce 100644
--- a/src/flb_engine.c
+++ b/src/flb_engine.c
@@ -604,7 +604,15 @@ int flb_engine_start(struct flb_config *config)
                 flb_sched_event_handler(config, event);
             }
             else if (event->type == FLB_ENGINE_EV_CUSTOM) {
-                event->handler(event);
+                if (event->status == MK_EVENT_REGISTERED) {
+                    flb_debug("[engine] calling registered callback for event: %i", event->fd);
+                    event->handler(event);
+                }
+                else {
+                    flb_debug("[engine] event is already de-registered: %p", event);
+                    /* still calling to see the core dump */
+                    event->handler(event);
+                }
             }
             else if (event->type == FLB_ENGINE_EV_THREAD) {
                 struct flb_upstream_conn *u_conn;
diff --git a/src/flb_upstream.c b/src/flb_upstream.c
index 1a967567..d0537a93 100644
--- a/src/flb_upstream.c
+++ b/src/flb_upstream.c
@@ -174,13 +174,20 @@ struct flb_upstream *flb_upstream_create_url(struct flb_config *config,
 
 static int destroy_conn(struct flb_upstream_conn *u_conn)
 {
+    int result = 0;
     struct flb_upstream *u = u_conn->u;
+    flb_debug("[upstream] backend: %s", mk_event_backend());
 
-    flb_trace("[upstream] destroy connection #%i to %s:%i",
-              u_conn->fd, u->tcp_host, u->tcp_port);
+    flb_debug("[upstream] destroy connection #%i, %p to %s:%i",
+              u_conn->fd, u_conn, u->tcp_host, u->tcp_port);
 
     if (u->flags & FLB_IO_ASYNC) {
-        mk_event_del(u->evl, &u_conn->event);
+        flb_debug("[upstream] destroy async connection #%i, %p to %s:%i -- %p, %p",
+                  u_conn->fd, u_conn, u->tcp_host, u->tcp_port, u->evl, &u_conn->event);
+        result = mk_event_del(u->evl, &u_conn->event);
+        printf("PHK event deleted %i\n", u_conn->fd);
+        flb_debug("[upstream] destroy async connection #%i, %p to %s:%i result : %i",
+                  u_conn->fd, u_conn, u->tcp_host, u->tcp_port, result);
     }
 
 #ifdef FLB_HAVE_TLS
@@ -191,7 +198,9 @@ static int destroy_conn(struct flb_upstream_conn *u_conn)
 #endif
 
     if (u_conn->fd > 0) {
-        u_conn->fd = -1;
+        flb_debug("[upstream] freeing async connection #%i, %p to %s:%i",
+                  u_conn->fd, u_conn, u->tcp_host, u->tcp_port);
+        //u_conn->fd = -1;
         flb_socket_close(u_conn->fd);
         /* remove connection from the queue */
         mk_list_del(&u_conn->_head);
@@ -324,7 +333,9 @@ struct flb_upstream_conn *flb_upstream_conn_get(struct flb_upstream *u)
 
     /* On non Keepalive mode, always create a new TCP connection */
     if (u->net.keepalive == FLB_FALSE) {
-        return create_conn(u);
+        conn = create_conn(u);
+        flb_debug("[upstream]keepalive false - created new connection, %p", conn);
+        return conn;
     }
 
     /*
@@ -333,6 +344,14 @@ struct flb_upstream_conn *flb_upstream_conn_get(struct flb_upstream *u)
      * entries exists, just create a new one.
      */
     mk_list_foreach_safe(head, tmp, &u->av_queue) {
+        /* In case we destroy the connection in the loop
+         * as is the case when KA connection is in failed state
+         * we will remove the element from the av_queue */
+        if (mk_list_is_empty(&u->av_queue) == 0) {
+            flb_debug("[upstream]no connections available");
+            conn = NULL;
+            break;
+        }
         conn = mk_list_entry(head, struct flb_upstream_conn, _head);
 
         /* This connection works, let's move it to the busy queue */
@@ -372,9 +391,11 @@ struct flb_upstream_conn *flb_upstream_conn_get(struct flb_upstream *u)
 
     /* No keepalive connection available, create a new one */
     if (!conn) {
+        flb_debug("[upstream]keepalive true - creating new connection");
         conn = create_conn(u);
     }
 
+    flb_debug("[upstream]keepalive true - created new connection, %p", conn);
     return conn;
 }
 
@@ -388,9 +409,11 @@ static int cb_upstream_conn_ka_dropped(void *data)
 
     conn = (struct flb_upstream_conn *) data;
 
-    flb_debug("[upstream] KA connection #%i to %s:%i has been disconnected "
+    flb_debug("[upstream] KA connection #%i, %p to %s:%i has been disconnected "
               "by the remote service",
-              conn->fd, conn->u->tcp_host, conn->u->tcp_port);
+              conn->fd, conn, conn->u->tcp_host, conn->u->tcp_port);
+
+    /* check if we still have the connection */
     return destroy_conn(conn);
 }
 
@@ -430,9 +453,12 @@ int flb_upstream_conn_release(struct flb_upstream_conn *conn)
             return destroy_conn(conn);
         }
 
-        flb_debug("[upstream] KA connection #%i to %s:%i is now available",
-                  conn->fd, conn->u->tcp_host, conn->u->tcp_port);
+        flb_debug("[upstream] KA connection #%i, %p to %s:%i -- %p, %p",
+                  conn->fd, conn, conn->u->tcp_host, conn->u->tcp_port, u->evl, &conn->event);
+
         conn->ka_count++;
+        flb_debug("[upstream] KA connection #%i, %p to %s:%i is now available (%i)",
+                  conn->fd, conn, conn->u->tcp_host, conn->u->tcp_port, conn->ka_count);
         return 0;
     }
```

Logs:
```
[1m[[0m2020/08/31 13:43:47[1m][0m [[93mdebug[0m] [storage] [cio file] synced at: mem.0/29883-1598861623.179194347.flb
[1m[[0m2020/08/31 13:43:47[1m][0m [[93mdebug[0m] [task] created task=0x5c2dda0 id=0 OK
[1m[[0m2020/08/31 13:43:47[1m][0m [[93mdebug[0m] [upstream]keepalive true - creating new connection
[1m[[0m2020/08/31 13:43:47[1m][0m [[93mdebug[0m] [upstream]keepalive true - created new connection, 0x5c342b0
[1m[[0m2020/08/31 13:44:24[1m][0m [[91merror[0m] [http_client] broken connection to 192.168.99.99:80 ?
[1m[[0m2020/08/31 13:44:24[1m][0m [[91merror[0m] [output:http:http.0] could not flush records to 192.168.99.99:80 (http_do=-1)
[1m[[0m2020/08/31 13:44:24[1m][0m [[93mdebug[0m] [upstream] KA connection #23, 0x5c342b0 to 192.168.99.99:80 -- 0x5c18180, 0x5c342b0
[1m[[0m2020/08/31 13:44:24[1m][0m [[93mdebug[0m] [upstream] KA connection #23, 0x5c342b0 to 192.168.99.99:80 is now available (1)
[1m[[0m2020/08/31 13:44:24[1m][0m [[94mtrace[0m] [input:mem:mem.0 at plugins/in_mem/mem.c:254] memory total=8168156 kb, used=7881708 kb, free=286448 kb
[1m[[0m2020/08/31 13:44:24[1m][0m [[94mtrace[0m] [input:mem:mem.0 at plugins/in_mem/mem.c:256] swap total=1942896 kb, used=1036 kb, free=1941860 kb
[1m[[0m2020/08/31 13:44:24[1m][0m [[93mdebug[0m] [storage] mem.0:29883-1598861664.963854203.flb mapped OK
[1m[[0m2020/08/31 13:44:24[1m][0m [[93mdebug[0m] [storage] [cio file] synced at: mem.0/29883-1598861664.963854203.flb
[1m[[0m2020/08/31 13:44:24[1m][0m [[93mdebug[0m] [task] created task=0x5c38710 id=1 OK
[1m[[0m2020/08/31 13:44:24[1m][0m [[93mdebug[0m] [upstream] KA connection #23 is in a failed state to: 192.168.99.99:80, cleaning up
[1m[[0m2020/08/31 13:44:24[1m][0m [[93mdebug[0m] [upstream] backend: epoll
[1m[[0m2020/08/31 13:44:24[1m][0m [[93mdebug[0m] [upstream] destroy connection #23, 0x5c342b0 to 192.168.99.99:80
[1m[[0m2020/08/31 13:44:24[1m][0m [[93mdebug[0m] [upstream] destroy async connection #23, 0x5c342b0 to 192.168.99.99:80 -- 0x5c18180, 0x5c342b0
[1m[[0m2020/08/31 13:44:24[1m][0m [[93mdebug[0m] [upstream] destroy async connection #23, 0x5c342b0 to 192.168.99.99:80 result : 0
[1m[[0m2020/08/31 13:44:24[1m][0m [[93mdebug[0m] [upstream] freeing async connection #23, 0x5c342b0 to 192.168.99.99:80
[1m[[0m2020/08/31 13:44:24[1m][0m [[93mdebug[0m] [upstream]keepalive true - creating new connection
[1m[[0m2020/08/31 13:44:24[1m][0m [[93mdebug[0m] [upstream]keepalive true - created new connection, 0x5c3efd0
[1m[[0m2020/08/31 13:45:02[1m][0m [[91merror[0m] [http_client] broken connection to 192.168.99.99:80 ?
[1m[[0m2020/08/31 13:45:02[1m][0m [[91merror[0m] [output:http:http.0] could not flush records to 192.168.99.99:80 (http_do=-1)
[1m[[0m2020/08/31 13:45:02[1m][0m [[93mdebug[0m] [upstream] KA connection #22, 0x5c3efd0 to 192.168.99.99:80 -- 0x5c18180, 0x5c3efd0
[1m[[0m2020/08/31 13:45:02[1m][0m [[93mdebug[0m] [upstream] KA connection #22, 0x5c3efd0 to 192.168.99.99:80 is now available (1)
[1m[[0m2020/08/31 13:45:02[1m][0m [[93mdebug[0m] [engine] event is already de-registered: 0x5c342b0
[1m[[0m2020/08/31 13:45:02[1m][0m [[93mdebug[0m] [upstream] KA connection #23, 0x5c342b0 to 192.168.99.99:80 has been disconnected by the remote service
[1m[[0m2020/08/31 13:45:02[1m][0m [[93mdebug[0m] [upstream] destroy connection #23, 0x5c342b0 to 192.168.99.99:80
[1m[[0m2020/08/31 13:45:02[1m][0m [[93mdebug[0m] [upstream] destroy async connection #23, 0x5c342b0 to 192.168.99.99:80 -- 0x5c18180, 0x5c342b0
[1m[[0m2020/08/31 13:45:02[1m][0m [[93mdebug[0m] [upstream] destroy async connection #23, 0x5c342b0 to 192.168.99.99:80 result : -1
[1m[[0m2020/08/31 13:45:02[1m][0m [[93mdebug[0m] [upstream] freeing async connection #23, 0x5c342b0 to 192.168.99.99:80
```